### PR TITLE
SAA-1572: add earliest release date to waiting list applications

### PR DESCRIPTION
**Pull request by dafhoustonmoj**
The /schedules/{scheduleId}/waiting-list-applications needs to include the earliest release date as this is required to be shown on the Waitlist on Activity Allocation screen

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/WaitingListService.kt [hunk 1]
```diff
@@ -56,7 +57,17 @@ class WaitingListService(
...
       .findOrThrowNotFound(id)
       .checkCaseloadAccess()
-      .let(waitingListRepository::findByActivitySchedule).map(WaitingList::toModel)
+      .let(waitingListRepository::findByActivitySchedule).map(WaitingList::toModel).apply {
+        val prisoners =
+          prisonerSearchApiClient.findByPrisonerNumbers(map { it.prisonerNumber })
+
+        map {
+          val prisoner = prisoners.find { p -> it.prisonerNumber == p.prisonerNumber }
+          if (prisoner != null) {
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/WaitingListService.kt:66]: Hmm `prisoner == null` should never happen.

If it does then we should probably fail fast here, something is wrong with our data.

Doing this though obviously has ramifications as retrieval of waiting lists will fail also.

Comment (dafhoustonmoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/WaitingListService.kt:66]: Does that mean the test I edited below (WaitingListServiceTest ) the line below could never happen? I should always return a value? If I added a return object it wouldn't need a null check. 
 
prisonerSearchApiClient.stub {
      on { findByPrisonerNumbers(listOf("123456")) } doReturn emptyList()
    }

The earliestReleaseDate is optional in the schema

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/model/WaitingListApplication.kt [hunk 2]
```diff
@@ -114,4 +115,7 @@ data class WaitingListApplication(
     example = "Jane Doe",
   )
   val updatedBy: String? = null,
+
+  @Schema(description = "The earliest release date of the prisoner")
+  var earliestReleaseDate: EarliestReleaseDate? = null,
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/model/WaitingListApplication.kt:120]: Will this ever be null?  It looks like it always returns something in the service ...

Comment (dafhoustonmoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/model/WaitingListApplication.kt]: updated to be mandatory

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/WaitingListService.kt [hunk 3]
```diff
@@ -49,14 +50,22 @@ class WaitingListService(
...
-      .let(waitingListRepository::findByActivitySchedule).map(WaitingList::toModel)
+
+    val prisoner =
+      prisonerSearchApiClient.findByPrisonerNumber(waitingList.prisonerNumber) ?: throw NullPointerException("Prisoner ${waitingList.prisonerNumber} not found for waiting list id $id")
+    return waitingList.toModel(determineEarliestReleaseDate(prisoner))
+  }
+
+  fun getWaitingListsBySchedule(id: Long) =
+    scheduleRepository.findOrThrowNotFound(id).checkCaseloadAccess().let(waitingListRepository::findByActivitySchedule).map {
+      val prisoner = prisonerSearchApiClient.findByPrisonerNumber(it.prisonerNumber) ?: throw NullPointerException("Prisoner ${it.prisonerNumber} not found for waiting list id $id")
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/WaitingListService.kt:65]: Sorry Daf.  I don't think this one is quite right.  Have you changed this to how it was in your previous commit?

This is going to make a prisoner-search request for each waiting list on the schedule.  So e.g. 100 waiting lists on a schedule then 100 external API requests.

So before I think you had this:

```
fun getWaitingListsBySchedule(id: Long) =
    scheduleRepository
      .findOrThrowNotFound(id)
      .checkCaseloadAccess()
      .let(waitingListRepository::findByActivitySchedule).map(WaitingList::toModel).apply {
        val prisoners =
          prisonerSearchApiClient.findByPrisonerNumbers(map { it.prisonerNumber })

        map {
          val prisoner = prisoners.find { p -> it.prisonerNumber == p.prisonerNumber } ?: throw NullPointerException("Prisoner ${it.prisonerNumber} not found for waiting list id $id")
          it.earliestReleaseDate = determineEarliestReleaseDate(prisoner)
        }
      }
```

**Review by matt-crowson-moj — APPROVED** [review]
LGTM - I presume this has a dependent UI PR which will need to be merged at the same time as this one.

**Pull request merged**